### PR TITLE
[bug] Set device on LightningModule after fitting

### PR DESCRIPTION
**Pull request by tchaton**
## What does this PR do?

This PR resolve a bug with device placement.
When running with gpu accelerator, the model was correctly put on cpu but the device was still set to cuda.

## Before submitting
- [ ] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/accelerators/gpu.py [hunk 1]
```diff
@@ -47,9 +47,9 @@ def on_train_start(self) -> None:
         with torch.cuda.device(self.root_device):
             torch.cuda.empty_cache()
 
-    def on_train_end(self) -> None:
+    def teardown(self) -> None:
+        self.lightning_module.device = torch.device("cpu")
         # clean up memory
-        self.model.cpu()
```
Comment (awaelchli) [inline pytorch_lightning/accelerators/gpu.py:52]: here, the model is wrapped in DistributedDataParallel
if we call `.cpu()` it will not call .cpu() on the modules and instead run the self.apply on the submodules for the parameters.
The fix we can do is call `self.lightning_module.cpu()`

File: pytorch_lightning/utilities/device_dtype_mixin.py [hunk 2]
```diff
@@ -48,9 +48,11 @@ def device(self) -> Union[str, torch.device]:
...
 
     @device.setter
-    def device(self, new_device: Union[str, torch.device]):
-        # Necessary to avoid infinite recursion
-        raise RuntimeError('Cannot set the device explicitly. Please use module.to(new_device).')
+    def device(self, device: Union[str, torch.device]):
+        if device.type == "xla":
+            raise MisconfigurationException("xla should be set using device")
+        self._device = device
+        self.to(device)
```
Comment (awaelchli) [inline pytorch_lightning/utilities/device_dtype_mixin.py:55]: not sure why this setter exists. can we delete it? 
I can't see it being used anywhere.

File: pytorch_lightning/accelerators/gpu.py (hunk 1 above)
Comment (tchaton) [reply pytorch_lightning/accelerators/gpu.py:52]: Done !

File: pytorch_lightning/utilities/device_dtype_mixin.py (hunk 2 above)
Comment (tchaton) [reply pytorch_lightning/utilities/device_dtype_mixin.py]: Done !

File: pytorch_lightning/utilities/device_dtype_mixin.py [hunk 3]
```diff
@@ -194,4 +189,8 @@ def apply_fn(module):
             if dtype is not None:
                 module._dtype = dtype
 
+        # update LightningModule device
+        self._device = device
```
Comment (awaelchli) [inline pytorch_lightning/utilities/device_dtype_mixin.py:193]: it's not necessary, as I see it `self.apply` already runs the apply_fn to itself.
https://github.com/pytorch/pytorch/blob/0a72904ab4a2eef45fab33b3b7ed79a9ebfa2d60/torch/nn/modules/module.py#L617
Correct?

Furthermore, there would probably be a bug here because device can be None!
If we do `self.half()` this would set the device to None! Careful

Comment (tchaton) [reply pytorch_lightning/utilities/device_dtype_mixin.py]: thanks, I didn't know that. Yes, it works without it as expected.

**Review by awaelchli — APPROVED** [review]
LGTM.
Minor nit: perhaps the test could have been integrated in an existing test that just runs fit, in one of the tests/models/test_gpu tests perhaps.
but minor

**Review by ethanwharris — APPROVED** [review]
LGTM :smiley:

File: tests/core/test_lightning_module.py [hunk 4]
```diff
@@ -358,3 +360,26 @@ def configure_optimizers(self):
...
+    def assert_device(device: torch.device) -> None:
+        assert model.device == device
+        for p in model.parameters():
+            assert p.device == device
+
+    assert_device(torch.device("cpu"))
+    model.to(torch.device("cuda:0"))
+    assert_device(torch.device("cuda:0"))
+
+    trainer = Trainer(default_root_dir=tmpdir, fast_dev_run=True, gpus=1)
```
Comment (carmocca) [inline tests/core/test_lightning_module.py:381]: Do we need to create a new trainer instance?

```suggestion
    trainer = Trainer(default_root_dir=tmpdir, fast_dev_run=True, gpus=1)
```

File: CHANGELOG.md [hunk 5]
```diff
@@ -322,6 +322,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Fixed parsing for pre-release package versions ([#6999](https://github.com/PyTorchLightning/pytorch-lightning/pull/6999))
 
 
+- Fixed resetting device after `fitting/testing/predicting` ([#7188](https://github.com/PyTorchLightning/pytorch-lightning/pull/7188))
```
Comment (carmocca) [inline CHANGELOG.md:325]: Validation got fixed too!

```suggestion
- Fixed resetting device after `fitting/evaluating/predicting` ([#7188](https://github.com/PyTorchLightning/pytorch-lightning/pull/7188))
```

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/accelerators/gpu.py [hunk 6]
```diff
@@ -47,9 +47,10 @@ def on_train_start(self) -> None:
         with torch.cuda.device(self.root_device):
             torch.cuda.empty_cache()
 
-    def on_train_end(self) -> None:
+    def teardown(self) -> None:
+        self.lightning_module.cpu()
```
Comment (ananthsub) [inline pytorch_lightning/accelerators/gpu.py:51]: @SeanNaren for sharded training, do we need to move both self.model and self.lightning_module? Could this lead to errors like what we saw for eval/train with SDP earlier?
